catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java

```diff
@@ -0,0 +1,175 @@
+import org.eclipse.jetty.http.HttpStatus;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import spark.Request;
+import spark.Response;
+import spark.Route;
+import spark.utils.IOUtils;
+
+public class CqlTransformHandler implements Route {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CqlTransformHandler.class);
+
+  private List<ServiceReference> queryResponseTransformers;
+
+  private BundleContext bundleContext;
+
+  private static final String GZIP = "gzip";
+
+  private ObjectMapper mapper =
+      new ObjectMapperImpl(
+          new JsonParserFactory().usePropertyOnly(),
+          new JsonSerializerFactory()
+              .includeEmpty()
+              .includeNulls()
+              .includeDefaultValues()
+              .setJsonFormatForDates(false)
+              .useAnnotations());
+
+  private EndpointUtil util;
+
+  public CqlTransformHandler(
+      List<ServiceReference> queryResponseTransformers,
+      BundleContext bundleContext,
+      EndpointUtil endpointUtil) {
+    this.queryResponseTransformers = queryResponseTransformers;
+    this.bundleContext = bundleContext;
+    this.util = endpointUtil;
+  }
+
+  @Override
+  public Object handle(Request request, Response response) throws Exception {
+    String transformerId = request.params(":transformerId");
+    String body = util.safeGetBody(request);
+    CqlRequest cqlRequest;
+
+    try {
+      cqlRequest = mapper.readValue(body, CqlRequest.class);
+    } catch (Exception e) {
+      LOGGER.debug("Error fetching cql request, empty or invalid body.");
+      response.status(HttpStatus.BAD_REQUEST_400);
+      return mapper.toJson(ImmutableMap.of("message", "Bad request"));
+    }
+
+    LOGGER.trace("Finding transformer to transform query response.");
+
+    ServiceReference<QueryResponseTransformer> queryResponseTransformer =
+        queryResponseTransformers
+            .stream()
+            .filter(transformer -> transformer.getProperty("id").equals(transformerId))
+            .findFirst()
+            .orElse(null);
+
+    if (queryResponseTransformer == null) {
+      LOGGER.debug("Could not find transformer with id: {}", transformerId);
+      response.status(HttpStatus.NOT_FOUND_404);
+      return mapper.toJson(ImmutableMap.of("message", "Service not found"));
+    }
+
+    CqlQueryResponse cqlQueryResponse = util.executeCqlQuery(cqlRequest);
+
+    attachFileToResponse(request, response, queryResponseTransformer, cqlQueryResponse);
+
+    return "";
+  }
+
+  private void setHttpHeaders(Request request, Response response, BinaryContent content)
+      throws MimeTypeException {
+    String mimeType = content.getMimeTypeValue();
+
+    if (mimeType == null) {
+      LOGGER.debug("Failure to fetch file extension, mime-type is empty");
+      throw new MimeTypeException("Binary Content contains null mime-type value.");
```

Comment:
What exceptions get thrown back to the user when we throw exceptions like this?

Reply:
Spark catches the exception and returns 500 with response "Error submitting feedback"